README.md
[![PSL tool](https://img.shields.io/badge/PSL-tool-a0a0a0.svg)](https://www.PSLmodels.org)
[![Python 3.6+](https://img.shields.io/badge/python-3.6%2B-blue.svg)](https://www.python.org/downloads/release/python-360/)


Using Package-Builder's pbrelease tool
======================================

This document tells you how to use the command-line interface (CLI) to
Package-Builder, which is called the `pbrelease` tool, to build conda
packages for your Policy Simulation Library (PSL) model and upload
them to the Anaconda Cloud's `PSLmodels` channel for public
distribution.  If you have questions about using `pbrelease`, or
experience problems, or want to request an enhancement, create a new
issue
[here](https://github.com/PSLmodels/Package-Builder/issues)
that poses your question, or provides details on what you think is
wrong with `pbrelease`, or describes the enhancement you're
requesting.


What is pbrelease?
------------------

The code in this Package-Builder GitHub repository, which is part of
the `PSLmodels` GitHub organization, generates a conda package called
`pkgbld` that contains the `pbrelease` tool.  For models in the
`PSLmodels` GitHub organization that meet certain criteria (see
below), `pbrelease` can be used to build conda packages (for Windows,
Linux, and Mac) for a specified model release version and then to
upload those built packages to the Anaconda Cloud's `PSLmodels`
channel for public distribution.

How to install pbrelease?
-------------------------

At the operating-system command prompt on a computer that has the
Anaconda distribution of Python 3.6 or 3.7 installed, execute this
command if you have never installed the package that contains the
`pbrelease` tool:

```
$ conda install -c PSLmodels pkgbld --yes
```

If you have previously installed that package, replace `install` with
`update` to upgrade to the most recent version of the package.

Next confirm that the installation/update went smoothly by executing
this command and getting something like this screen output:

```
$ pbrelease --version
Package-Builder 0.21.0
```

Then see the information that `pbrelease` expects by asking for help
and getting something like this screen output:

```
$ pbrelease --help
usage: pbrelease  REPOSITORY_NAME  PACKAGE_NAME  MODEL_VERSION
                  [--help]  [--local]  [--dryrun]  [--version]

Creates conda packages named PACKAGE_NAME for the PSL model in REPOSITORY_NAME
that has a GitHub release named MODEL_VERSION. The packages are built locally
in a temporary workspace and then uploaded to the Anaconda Cloud PSLmodels
channel for public distribution. The built/uploaded packages are for Python
3.6 and Python 3.7.

positional arguments:
  REPOSITORY_NAME  Name of repository in the GitHub organization called
                   PSLmodels. Example: Tax-Calculator
  PACKAGE_NAME     Name of packages to build and upload. Example: taxcalc
  MODEL_VERSION    Model release string that has X.Y.Z semantic-versioning
                   pattern. Example: 1.0.1

optional arguments:
  -h, --help       show this help message and exit
  --local          optional flag that causes package to be built from current
                   source code and installed on local computer without
                   packages being uploaded to Anaconda Cloud PSLmodels
                   channel.
  --dryrun         optional flag that writes execution plan to stdout and
                   quits without executing plan
  --version        optional flag that writes Package-Builder release version
                   to stdout and quits
```

Once you can get this kind of `pbrelease` screen output, you're ready
to make packages for your model release.


How to use pbrelease?
---------------------

In order to use the `pbrelease` tool, you need a file containing the
Anaconda Cloud upload token for the `PSLmodels` channel.  Get that
file by asking for it in a new Package-Builder issue.  You will be
sent via private email a zip file containing the token file.  Unzip
that file in your home directory.  After doing that, the
`.pslmodels_anaconda_token` file should be in your home directory.
Now you're ready to use the `pbrelease` tool.

After testing your model and commiting to the master branch all of
your changes, create a model release on GitHub.  Then to build and
upload packages for that model release, execute this command:

```
$ pbrelease  repo_name  pkg_name  version
```

where your replace `repo_name` with the name of your model's
repository and you replace `pkg_name` with the sub-directory name
containing your model's highest level `__init__.py` file and you
replace `version` with the release for which you want to make
model packages.

Here's an example that builds and uploads packages for
Behavioral-Responses release 0.5.0:

```
$ pbrelease Behavioral-Responses behresp 0.5.0
: Package-Builder will build model packages for:
:   repository_name = Behavioral-Responses
:   package_name = behresp
:   model_version = 0.5.0
:   python_versions = ['3.6', '3.7']
: Package-Builder will upload model packages to:
:   Anaconda channel = pslmodels
:   using token in file = /Users/mrh/.pslmodels_anaconda_token
: Package-Builder is starting at Thu Mar 14 07:25:21 2019
: Package-Builder is cloning repository code for 0.5.0
...
... <an enormous amount of screen output>
...
: Package-Builder is cleaning-up
: Package-Builder is finishing at Thu Mar 14 07:27:27 2019
```

To use `pbrelease` to make a **local package** from the current (even
uncommitted) code on your current branch, make the top-level directory
of the repo's source-code tree the current working directory.  Then
use the `--local` option as shown here:

```
$ cd Tax-Calculator
$ pbrelease Tax-Calculator taxcalc 0.0.0 --local
```

This will produce a local package on your computer for testing or
validation work.  The above example specifies version 0.0.0, but you
can specify any version you want.  You can uninstall this local
package at any time using this command:

```
$ conda uninstall taxcalc --yes
```

What are the package-building criteria?
---------------------------------------

The `pbrelease` tool makes certain assumptions about the PSL
model's source code.  These assumptions must be met for `pbrelease` to
work correctly.  Here are the model code criteria:

(1) Model code must be compatible with Anaconda's Python 3.6 and higher

(2) Model code must be organized as follows:

```
<repo_name>               <--- TOP-LEVEL DIRECTORY (e.g., Tax-Calculator)
    setup.py
    ...
    conda.recipe          <--- SUB-DIRECTORY
        meta.yaml
        build.sh
        bld.bat
        ...
    <pkg_name>            <--- SUB-DIRECTORY (e.g., taxcalc)
        __init__.py
        ...
```

(3) The `setup.py` file must contain a line like this: `version = '0.0.0'`

(4) The `meta.yaml` file must contain a line like this: `version: 0.0.0`

(5) The `__init__.py` file must contain a line like this: `__version__ = '0.0.0'`

The best place to look at code that meets all these criteria is the
PSLmodels Behavioral-Responses repository.

Version 0.0.0 indicates the earliest version; the `pbrelease` tool will
automatically replace (in a temporary copy of your model code) the
0.0.0 with the X.Y.Z version specified when you execute `pbrelease`.


RELEASES.md
PACKAGE-BUILDER RELEASE HISTORY
===============================
Go [here](https://github.com/PSLmodels~/Package-Builder/pulls?q=is%3Apr+is%3Aclosed)
for a complete commit history.


2019-04-15 Release 0.22.0
-------------------------
(last merged pull request is
[#164](https://github.com/PSLmodels/Package-Builder/pull/164))

**API Changes**
- None

**New Features**
- Simplify `--local` so that it is a simple boolean option and allow the locally installed package to be specified with any version
  [[#164](https://github.com/PSLmodels/Package-Builder/pull/164)
  by Martin Holmer]

**Bug Fixes**
- None

2019-03-13 Release 0.21.0
-------------------------
(last merged pull request is
[#161](https://github.com/PSLmodels/Package-Builder/pull/161))

**API Changes**
- None

**New Features**
- None

**Bug Fixes**
- Correct name of directory where built package is to be found for local installation using the new `--local LOCAL` option
  [[#161](https://github.com/PSLmodels/Package-Builder/pull/161)
  by Martin Holmer]


2019-03-13 Release 0.20.0
-------------------------
(last merged pull request is
[#158](https://github.com/PSLmodels/Package-Builder/pull/158))

**API Changes**
- Remove `--also37` option as now packages for uploading are automatically built for both Python 3.6 and Python 3.7
  [[#158](https://github.com/PSLmodels/Package-Builder/pull/158)
  by Martin Holmer]

**New Features**
- Add `--local LOCAL` option to build package from local source code and install it on local computer
  [[#158](https://github.com/PSLmodels/Package-Builder/pull/158)
  by Martin Holmer with suggestion and assistance from Hank Doupe]

**Bug Fixes**
- None


2018-12-19 Release 0.19.0
-------------------------
(last merged pull request is
[#153](https://github.com/PSLmodels/Package-Builder/pull/153))

**API Changes**
- None

**New Features**
- Add datetime to :starting and :finishing prints
  [[#153](https://github.com/PSLmodels/Package-Builder/pull/153)
  by Martin Holmer]

**Bug Fixes**
- None


2018-12-18 Release 0.18.0
-------------------------
(last merged pull request is
[#150](https://github.com/PSLmodels/Package-Builder/pull/150))

**API Changes**
- None

**New Features**
- Use newer `subprocess.run` in `os_call` utility function
  [[#150](https://github.com/PSLmodels/Package-Builder/pull/150)
  by Martin Holmer]

**Bug Fixes**
- None


2018-12-17 Release 0.17.0
-------------------------
(last merged pull request is
[#149](https://github.com/PSLmodels/Package-Builder/pull/149))

**API Changes**
- None

**New Features**
- Add `anaconda upload` command print statement
  [[#149](https://github.com/PSLmodels/Package-Builder/pull/149)
  by Martin Holmer based on suggestion from Hank Doupe]

**Bug Fixes**
- None


2018-12-16 Release 0.16.0
-------------------------
(last merged pull request is
[#148](https://github.com/PSLmodels/Package-Builder/pull/148))

**API Changes**
- None

**New Features**
- Make `anaconda upload` command overwrite already uploaded packages
  [[#148](https://github.com/PSLmodels/Package-Builder/pull/148)
  by Martin Holmer with assistance from Hank Doupe]
- Avoid using the new 1.7.* versions of anaconda-client package
  [[#148](https://github.com/PSLmodels/Package-Builder/pull/148)
  by Martin Holmer with assistance from Hank Doupe]

**Bug Fixes**
- None


2018-12-15 Release 0.15.0
-------------------------
(last merged pull request is
[#147](https://github.com/PSLmodels/Package-Builder/pull/147))

**API Changes**
- None

**New Features**
- Simplify `anaconda upload` command used by Package-Builder
  [[#147](https://github.com/PSLmodels/Package-Builder/pull/147)
  by Martin Holmer with assistance from Hank Doupe]

**Bug Fixes**
- None


2018-12-15 Release 0.14.0
-------------------------
(last merged pull request is
[#146](https://github.com/PSLmodels/Package-Builder/pull/146))

**API Changes**
- None

**New Features**
- Add `anaconda-client` as a required dependency of Package-Builder
  [[#146](https://github.com/PSLmodels/Package-Builder/pull/146)
  by Martin Holmer with need pointed out by Hank Doupe]

**Bug Fixes**
- None


2018-12-14 Release 0.13.0
-------------------------
(last merged pull request is
[#144](https://github.com/PSLmodels/Package-Builder/pull/144))

**API Changes**
- Add `pbrelease` options `--also37` and `--dryrun` so that packages for Python 3.7 can also be built and uploaded, and so that the build/upload plan can be viewed without executing the plan
  [[#136](https://github.com/PSLmodels/Package-Builder/pull/136)
  by Martin Holmer]

**New Features**
- None

**Bug Fixes**
- None


2018-11-14 Release 0.12.1
-------------------------
(last merged pull request is
[#126](https://github.com/PSLmodels/Package-Builder/pull/126))

**API Changes**
- None

**New Features**
- Add easier-to-understand error message to `pbrelease` CLI if Anaconda token file is not present
  [[#125](https://github.com/PSLmodels/Package-Builder/pull/125)
  by Martin Holmer]
- Make `git clone` faster by downloading only repository code for the version being released (rather than the whole repository)
  [[#126](https://github.com/PSLmodels/Package-Builder/pull/126)
  by Martin Holmer]

**Bug Fixes**
- None


2018-11-13 Release 0.12.0
-------------------------
(last merged pull request is
[#123](https://github.com/PSLmodels/Package-Builder/pull/123))

**API Changes**
- Change `pbrelease` CLI error handling to show simpler error messages and help
  [[#123](https://github.com/PSLmodels/Package-Builder/pull/123)
  by Martin Holmer]

**New Features**
- None

**Bug Fixes**
- None


2018-11-13 Release 0.11.0
-------------------------
(last merged pull request is
[#120](https://github.com/PSLmodels/Package-Builder/pull/120))

**API Changes**
- Change approach to specifying the package version
  [[#120](https://github.com/PSLmodels/Package-Builder/pull/120)
  by Martin Holmer]

**New Features**
- None

**Bug Fixes**
- None


2018-11-12 Release 0.10.0
-------------------------

pkgbld/cli.py
"""
Command-line interface (CLI) to Package-Builder response function,
which can be accessed as 'pbrelease' from an installed pkgbld conda package.
"""
# CODING-STYLE CHECKS:
# pycodestyle cli.py
# pylint --disable=locally-disabled cli.py

import argparse
import re
import os
import pkgbld


def main():
    """
    Contains command-line interface (CLI) to Package-Builder response function.
    """
    # parse command-line arguments:
    usage_str = ('pbrelease  REPOSITORY_NAME  PACKAGE_NAME  MODEL_VERSION\n'
                 '                  [--help]  [--local]  '
                 '[--dryrun]  [--version]')
    parser = argparse.ArgumentParser(
        prog='',
        usage=usage_str,
        description=('Creates conda packages named PACKAGE_NAME for the PSL '
                     'model in REPOSITORY_NAME that has a GitHub release '
                     'named MODEL_VERSION.  The packages are built locally '
                     'in a temporary workspace and then uploaded to the '
                     'Anaconda Cloud PSLmodels channel for public '
                     'distribution.  The built/uploaded packages are '
                     'for Python 3.6 and Python 3.7.')
    )
    parser.add_argument('REPOSITORY_NAME', nargs='?',
                        help=('Name of repository in the GitHub organization '
                              'called PSLmodels. Example: Tax-Calculator'),
                        default=None)
    parser.add_argument('PACKAGE_NAME', nargs='?',
                        help=('Name of packages to build and upload. '
                              'Example: taxcalc'),
                        default=None)
    parser.add_argument('MODEL_VERSION', nargs='?',
                        help=('Model release string that has X.Y.Z '
                              'semantic-versioning pattern. '
                              'Example: 1.0.1'),
                        default=None)
    parser.add_argument('--local',
                        help=('optional flag that causes package to be '
                              'built from current source code and installed '
                              'on local computer without packages being '
                              'uploaded to Anaconda Cloud PSLmodels channel.'),
                        default=False,
                        action="store_true")
    parser.add_argument('--dryrun',
                        help=('optional flag that writes execution plan '
                              'to stdout and quits without executing plan'),
                        default=False,
                        action="store_true")
    parser.add_argument('--version',
                        help=('optional flag that writes Package-Builder '
                              'release version to stdout and quits'),
                        default=False,
                        action="store_true")
    args = parser.parse_args()
    # show Package-Builder version and quit if --version option specified
    if args.version:
        print('Package-Builder {}'.format(pkgbld.__version__))
        return 0
    # check command-line arguments
    repo_name = args.REPOSITORY_NAME
    pkg_name = args.PACKAGE_NAME
    version = args.MODEL_VERSION
    emsg = ''
    if not isinstance(repo_name, str):
        emsg += 'ERROR: REPOSITORY_NAME is not specified\n'
    if not isinstance(pkg_name, str):
        emsg += 'ERROR: PACKAGE_NAME is not specified\n'
    if not isinstance(version, str):
        emsg += 'ERROR: MODEL_VERSION is not specified\n'
    else:
        pattern = r'^[0-9]+\.[0-9]+\.[0-9]+$'
        if re.match(pattern, version) is None:
            emsg += ('ERROR: MODEL_VERSION does not have X.Y.Z '
                     'semantic-versioning pattern\n')
    if not os.path.isfile(pkgbld.ANACONDA_TOKEN_FILE):
        emsg += ('ERROR: Anaconda token file {} '
                 'does not exist\n'.format(pkgbld.ANACONDA_TOKEN_FILE))
    if args.local:
        cwd = os.getcwd()
        if not cwd.endswith(repo_name):
            emsg += ('ERROR: cwd={} does not correspond to '
                     'REPOSITORY_NAME={}\n'.format(cwd, repo_name))
        local_pkgname = os.path.join('.', pkg_name)
        if not os.path.isdir(local_pkgname):
            emsg += ('ERROR: cwd={} does not contain '
                     'subdirectory {} '.format(cwd, pkg_name))
    if emsg:
        print(emsg)
        print('USAGE:', usage_str)
        return 1
    # call pkgbld release function with specified parameters
    pkgbld.release(repo_name, pkg_name, version,
                   local=args.local, dryrun=args.dryrun)
    return 0
